src/main/kotlin/Main.kt

```diff
@@ -2,98 +2,192 @@ import com.beust.klaxon.JsonArray
+    val numTests = tests.size
+    val progress = AtomicInteger(0)
+    val failedTests = AtomicInteger(0)
+    println()
+    println("Running $numTests tests on engine \"${engine.name}\"")
 
-        val executor = Executor(engine)
+    val executor = Executor(engine)
 
-        val t = printProgressbar(progress, failed_tests, numTests)
+    val t = printProgressbar(progress, failedTests, numTests)
 
-        val time = measureTimeMillis {
-            tests.parallelStream().forEach {
+    val time = measureTimeMillis {
+        tests.parallelStream().forEach {
 
-                val result = executor.runTest(it)
+            val result = executor.runTest(it, engine.deadline)
 
-                results.add(ResultContext(result, engine.name, engine.version))
+            results.add(ResultContext(result, engine.name, engine.version))
 
-                progress.getAndAdd(1)
+            progress.getAndAdd(1)
 
-                if (result.result != result.expected) {
-                    if (engine.verbose == true) {
-                        print("\r") // Replace the progress bar
-                        printTestResult(result)
-                    }
-                    failed_tests.getAndAdd(1)
+            if (result.result != result.expected) {
+                if (engine.verbose == true) {
+                    print("\r") // Replace the progress bar
+                    printTestResult(result)
                 }
+                failedTests.getAndAdd(1)
             }
-            engine.terminate()
         }
+        engine.terminate()
+    }
 
-        t.join()
+    t.join()
 
-        var passed = 0
-        var failed = 0
-        var unknown = 0
+    var passed = 0
+    var failed = 0
+    var unknown = 0
 
-        results.forEach { when(it.result.result) {
-            it.result.expected ->  passed++
+    results.forEach {
+        when (it.result.result) {
+            it.result.expected -> passed++
             else -> if (it.result.result == ResultType.EXCEPTION) unknown++ else failed++
-            }
         }
+    }
 
-        println()
-        println("${passed}/$numTests tests succeeded (${(passed*100.0/numTests).roundToInt()}%)" +
+    println()
+    println(
+        "${passed}/$numTests tests succeeded (${(passed * 100.0 / numTests).roundToInt()}%)" +
                 (if (unknown == 0) "" else " ($unknown failed due to exceptions)") +
-                " in ${time/1000} seconds")
+                " in ${time / 1000} seconds"
+    )
+    return results
+}
+
+private fun generateTests(): Collection<Test> {
+    val allRelations = ProofSearcher().addAllProofs().findNewRelations(RelationLoader.relations)
+    //ProofLimiter(3).limit(allRelations)
+    return TestGenerator().addAllTests().generateTests(allRelations)
+}
+
+private fun sortTests(engine: EngineConfiguration, tests: Collection<Test>) : Collection<Test> {
+    val operators = listOf("||", "\\\\", "&&", "consistency:", "refinement:")
```

Comment:
Are queries with multiple refinements/consistency checks generated? If this is the case, what is the complexity of this query: 
``refinement:Spec<=(Administration||Machine||Researcher); refinement:Machine3<=Machine3`` has complexity: "refinement:" + "<=" + "||" + "||" + "refinement:" + "<=" = 6
In this case, I believe the complexity should be calculated for each of the refinements:
``refinement:Spec<=(Administration||Machine||Researcher);`` has complexity: "refinement:" + "<=" + "||" + "||" = 3
``refinement:Machine3<=Machine3`` has complexity: "refinement:" + "<=" = 1

Reply:
Based on some quick testing, it does not generate these queries

Reply:
Fantastic :+1: